### DISCUSSION TITLE
How to get repository watchers count

### DISCUSSION THREAD

**Discussion by brunoLPZ** [octokit/octokit.js#2140] [Q&A]
I want to get the repository watchers count without having to call the `listWatchersForRepo` method. I'm using the `listReposAccessibleToInstallation` to get all repositories and I was expecting to get the watchers count from the `watchers` or `watchers_count` field. However this value doesn't seem to be correct. I've also was trying to use the `subscribers_count` field but it isn't available. If I use the github API directly I can see the correct count in `subscribers_count`. 

I'm using the following dependencies:

```
"dependencies": {
    "@octokit/rest": "^18.6.7",
    "@octokit/auth-app": "^3.5.3"
  },
  "devDependencies": {
    "@octokit/types": "^6.19.0"
  }
```

**Comment by gr2m** [comment]
I think `watchers_count` is deprecated, it has been renamed to to `stargazers_count`. The names have been changed in the past in GitHub's UI, but APIs cannot change without an explicit new version, so here we are. `subscribers_count` is what you are looking for.

I was able to confirm that `.subscribers_count` is not est when using the [`GET /installation/repositories` (List repositories accessible to the app installation)](https://docs.github.com/en/rest/reference/apps#list-repositories-accessible-to-the-app-installation) endpoint. In the OpenAPI Schema, the `.subscribers_count` response key is defined, but not required. But I don't know under what circumstances it is set. I tried registering an app with all permissions enabled, but the `.subscribers_count` was still not set.

At this point I suggest you reach out to support at https://support.github.com/contact, you can reference this issue. This might be a legit bug, and if not they will at least be able to explain the current behavior.

As an alternative to your problem, you can use GraphQL

```js
const { view: { repositories: { nodes: repos } } } = await octokit.graphql(`{
  viewer {
    repositories(first: 100) {
      nodes {
        name
        watchers(first: 1) {
          totalCount
        }
      }
    }
  }
}`)

console.log(repos)
```

The challenge here is that you can only get the first 100, after that you need to paginate the GraphQL query, which is a bit more complicated than with the REST API. I can give you some pointers if you need them

**Reply by brunoLPZ** [reply]
For now I've just used the [GET /repos/{owner}/{repo}/subscribers](https://docs.github.com/en/rest/reference/activity#list-watchers) endpoint applying pagination with a single item per page. In this way, I just need to perform a single request that gets only one item. Based on the response headers I can determine the number of watchers as there is a `link` header that should contain something like `<https://api.github.com/...?page=10&per_page=1>; rel="last"`. Parsing this gives me the last page number that should be the same as the watchers count.

Anyway, thanks for your clarification and suggestion @gr2m, I'll address this question to the github support.

**Reply by gr2m** [reply]
if you only use the response headers, you can use 

```js
const { headers } = await octokit.request("HEAD /repos/{owner}/{repo}/subscribers", { owner, repo, per_page: 1 })
```

Important caveat: the `link` header **is not set** if there is no 2nd page. So if the repository has 0 or 1 subscribers, your code might throw if you rely on the `link` header to be present.

**Reply by brunoLPZ** [reply]
Yes I've already taken into account the case where there's just 0 or 1 subscribers. But great point about the `HEAD` method. Thanks a lot!

**Comment by tik9** [comment]
@gr2m I tried your example from 23 july last year and get an undefined. This is what I have, can you help:
```js
const graphqlAuth = graphql.defaults({headers: { authorization: 'token ' + process.env.ghtoken, },});});
const { repos } = await graphqlAuth(`{repositoryOwner(login: "tik9"){repositories(first: 2) {nodes {name watchers(first: 2) {totalCount}}}}}`)
console.log(repos)
```

**Comment by gr2m** [comment]
your code is incorrect, you are destructing `repos` from the result that `graphqlAuth` resolves to. But as you see in my code, it's

```js
const { view: { repositories: { nodes: repos } } } = await ...
```
